pkg/abstractions/endpoint/buffer.go

```diff
@@ -232,43 +236,71 @@ func (rb *RequestBuffer) discoverContainers() {
 			rb.availableContainers = availableContainers
 			rb.availableContainersLock.Unlock()
 
-			time.Sleep(1 * time.Second)
+			time.Sleep(500 * time.Millisecond)
 		}
 	}
 }
 
-func (rb *RequestBuffer) requestsInFlight(containerId string) (int, error) {
-	val, err := rb.rdb.Get(rb.ctx, Keys.endpointRequestsInFlight(rb.workspace.Name, rb.stubId, containerId)).Int()
+func (rb *RequestBuffer) requestTokens(containerId string) (int, error) {
+	tokenKey := Keys.endpointRequestTokens(rb.workspace.Name, rb.stubId, containerId)
+
+	val, err := rb.rdb.Get(rb.ctx, tokenKey).Int()
 	if err != nil && err != redis.Nil {
 		return 0, err
 	} else if err == redis.Nil {
-		return 0, nil
+		maxTokens := int(rb.stubConfig.Workers)
+
+		created, err := rb.rdb.SetNX(rb.ctx, tokenKey, maxTokens, 0).Result()
+		if err != nil {
+			return 0, err
+		}
+
+		if created {
+			return maxTokens, nil
+		}
+
+		tokens, err := rb.rdb.Get(rb.ctx, tokenKey).Int()
+		if err != nil {
+			return 0, err
+		}
+
+		return tokens, nil
 	}
 
 	return val, nil
 }
 
-func (rb *RequestBuffer) incrementRequestsInFlight(containerId string) error {
-	err := rb.rdb.Incr(rb.ctx, Keys.endpointRequestsInFlight(rb.workspace.Name, rb.stubId, containerId)).Err()
+func (rb *RequestBuffer) acquireRequestToken(containerId string) error {
+	tokenKey := Keys.endpointRequestTokens(rb.workspace.Name, rb.stubId, containerId)
+	tokenCount, err := rb.rdb.Decr(rb.ctx, tokenKey).Result()
 	if err != nil {
 		return err
 	}
 
-	err = rb.rdb.Expire(rb.ctx, Keys.endpointRequestsInFlight(rb.workspace.Name, rb.stubId, containerId), time.Duration(rb.stubConfig.TaskPolicy.Timeout)*time.Second).Err()
+	// If the token count is negative, we exceeded our threshold of
+	// available request tokens, just reverse the operation
+	if tokenCount < 0 {
+		rb.rdb.Incr(rb.ctx, tokenKey)
+		return errors.New("too many in-flight requests")
+	}
+
+	err = rb.rdb.Expire(rb.ctx, tokenKey, time.Duration(rb.stubConfig.TaskPolicy.Timeout)*time.Second).Err()
```

Comment:
Let's say 2 gateways are running: G1 and G2. G1 has 3/5 tokens and G2 has 2/5 tokens. 
If G1 crashes. 3 tokens are just gone. Since G2 keeps refreshing the expiration, the key never expires and resets

Reply:
The tokens are by container, not by gateway - but yes it's true that if a gateway crashes, it wouldn't decrement the tokens so they would hang around until the expiration